src/Database/LSMTree/Internal/WriteBufferWriter.hs

```diff
@@ -0,0 +1,227 @@
+import           Control.Monad.Primitive (PrimMonad (..))
+import           Control.Monad.ST (ST)
+import           Data.Foldable (for_)
+import           Data.Maybe (maybeToList)
+import           Data.Primitive.PrimVar (PrimVar, newPrimVar)
+import           Data.Word (Word64)
+import           Database.LSMTree.Internal.BlobFile (BlobSpan)
+import           Database.LSMTree.Internal.BlobRef (RawBlobRef)
+import           Database.LSMTree.Internal.ChecksumHandle (ChecksumHandle,
+                     closeHandle, copyBlob, dropCache, makeHandle, readChecksum,
+                     writeRawOverflowPages, writeRawPage)
+import qualified Database.LSMTree.Internal.CRC32C as CRC
+import           Database.LSMTree.Internal.Entry (Entry)
+import           Database.LSMTree.Internal.PageAcc (PageAcc)
+import qualified Database.LSMTree.Internal.PageAcc as PageAcc
+import qualified Database.LSMTree.Internal.PageAcc1 as PageAcc
+import           Database.LSMTree.Internal.Paths (ForWriteBufferFiles (..),
+                     WriteBufferFsPaths, forWriteBufferBlobRaw,
+                     forWriteBufferKOpsRaw, pathsForWriteBufferFiles,
+                     toChecksumsFileForWriteBufferFiles,
+                     writeBufferChecksumsPath)
+import           Database.LSMTree.Internal.RawOverflowPage (RawOverflowPage)
+import           Database.LSMTree.Internal.RawPage (RawPage)
+import           Database.LSMTree.Internal.Serialise (SerialisedKey,
+                     SerialisedValue)
+import           Database.LSMTree.Internal.WriteBuffer (WriteBuffer)
+import qualified Database.LSMTree.Internal.WriteBuffer as WB
+import           Database.LSMTree.Internal.WriteBufferBlobs (WriteBufferBlobs)
+import qualified Database.LSMTree.Internal.WriteBufferBlobs as WBB
+import qualified System.FS.API as FS
+import           System.FS.API (HasFS)
+import qualified System.FS.BlockIO.API as FS
+import           System.FS.BlockIO.API (HasBlockIO)
+
+
+writeWriteBuffer ::
+     (MonadSTM m, MonadST m, MonadThrow m)
+  => HasFS m h
+  -> HasBlockIO m h
+  -> WriteBufferFsPaths
+  -> WriteBuffer
+  -> WriteBufferBlobs m h
+  -> m ()
+writeWriteBuffer hfs hbio fsPaths buffer blobs = do
+  writer <- new hfs hbio fsPaths
+  for_ (WB.toList buffer) $ \(key, op) ->
+    -- TODO: The fmap entry here reallocates even when there are no blobs.
+    addKeyOp writer key (WBB.mkRawBlobRef blobs <$> op)
+  void $ unsafeFinalise True writer
+
+-- data SerialisedWriteBuffer m h = SerialisedWriteBuffer
+--     { -- | The file system paths for all the files used by the serialised write buffer.
+--       serialisedWriteBufferFsPaths    :: !WriteBufferFsPaths
+--       -- | The (read mode) file handles.
+--     , serialisedWriteBufferHandles    :: !(ForWriteBufferFiles (FS.Handle h))
+--     , serialisedWriteBufferHasFS      :: !(HasFS m h)
+--     , serialisedWriteBufferHasBlockIO :: !(HasBlockIO m h)
+--     }
+
+data WriteBufferWriter m h = WriteBufferWriter
```

Comment:
Give this some haddocks. Maybe also hint at the fact that the `WriteBufferWriter` is largely based on `RunBuilder`, and that is why we link its documenation some places.